File: model-client/src/main/java/org/modelix/model/util/StreamUtils.java
package org.modelix.model.util;

import java.util.List;
import java.util.PrimitiveIterator;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.LongStream;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

public class StreamUtils {
    public static <T> Stream<T> toStream(Iterable<T> iterable) {
        return StreamSupport.stream(iterable.spliterator(), false);
    }

    public static int indexOf(LongStream stream, long value) {
        PrimitiveIterator.OfLong iterator = stream.iterator();
        int index = 0;
        while (iterator.hasNext()) {
            if (iterator.nextLong() == value) {
                return index;
            }
            index++;
        }
        return -1;
    }

    public static <T> Set<T> intersection(Stream<T> a, Set<T> b) {
        return a.filter(b::contains).collect(Collectors.toSet());
    }

    public static <T> T last(List<T> list) {
        return list.isEmpty() ? null : list.get(list.size() - 1);
    }

    public static <T> T removeLast(List<T> list) {
        return list.isEmpty() ? null : list.remove(list.size() - 1);
    }
}

File: model-client/src/main/java/org/modelix/model/util/StreamUtils.kt
package org.modelix.model.util

import java.util.stream.Collectors
import java.util.stream.LongStream
import java.util.stream.Stream
import java.util.stream.StreamSupport

object StreamUtils {
    fun <T> toStream(iterable: Iterable<T>): Stream<T> {
        return StreamSupport.stream(iterable.spliterator(), false)
    }

    fun indexOf(stream: LongStream, value: Long): Int {
        val iterator = stream.iterator()
        var index = 0
        while (iterator.hasNext()) {
            if (iterator.nextLong() == value) {
                return index
            }
            index++
        }
        return -1
    }

    fun <T> intersection(a: Stream<T>, b: Set<T>): Set<T> {
        return a.filter { o: T -> b.contains(o) }.collect(Collectors.toSet())
    }

    fun <T> last(list: List<T>): T? {
        return if (list.isEmpty()) null else list[list.size - 1]
    }

    fun <T> removeLast(list: MutableList<T>): T? {
        return if (list.isEmpty()) null else list.removeAt(list.size - 1)
    }
}